### DISCUSSION TITLE
Cannot spawn and maintain position of multiple models

### DISCUSSION THREAD

**Discussion by leeenglestone** [StereoKit/StereoKit#391] [Q&A] [answered]
I have looked at the various demos including DemoHands and DemoManyObjects and must be doing something wrong..

I am trying to use the hand menu to spawn a new instance of a model and want to be able to manipulate each models position independently.. I seem to be adding new instances of the Model to the app but they all want to be at the same positioning when moving..

I've even tried maintaining a separate list of Models but I don't think that is needed..

`
using StereoKit;
using System.Collections.Generic;
using System.Numerics;

namespace StereoKitApp
{
    public class App
    {
        public SKSettings Settings => new SKSettings
        {
            appName = "StereoKit Template",
            assetsFolder = "Assets",
            displayPreference = DisplayMode.MixedReality
        };

        bool showHandMenus = true;
       
        Matrix4x4 floorTransform = Matrix.TS(new Vector3(0, -1.5f, 0), new Vector3(30, 0.1f, 30));
        Material floorMaterial;

        // Maybe still too far away?
        Pose kanbanBoardPose = new Pose(0f, 0f, -3f, Quat.LookDir(0, 0, 0));

        static List<Pose> poseCache = new List<Pose>();

        Pose windowPose2 = new Pose(-0.5f, 0, -0.5f, Quat.LookDir(-1, 0, 1));

        Model model;

        public void Init()
        {
            // Create assets used by the app           
            floorMaterial = new Material(Shader.FromFile("floor.hlsl"));
            floorMaterial.Transparency = Transparency.Blend;

            model = Model.FromMesh(
                    Mesh.GenerateRoundedCube(Vec3.One * 0.1f, 0.02f),
                    Default.MaterialUI);

            poseCache.Add(new Pose(0, 0, -0.5f, Quat.Identity));
            //models.Add(model);
        }

        public void Step()
        {
            if (SK.System.displayType == Display.Opaque)
                Default.MeshCube.Draw(floorMaterial, floorTransform);                                   

            if (showHandMenus)
            {
                DrawHandMenu(Handed.Right);
                DrawHandMenu(Handed.Left);
            }

            //var model = models[0];
            Pose curr = poseCache[0];
            UI.Handle("Cube", ref curr, model.Bounds);

            for (int x = 0; x < poseCache.Count; x++)
            {
                //model = models[x];
                Pose pose = poseCache[x];
                model.Draw(pose.ToMatrix());
            }

            poseCache.RemoveAt(poseCache.Count - 1);
            poseCache.Insert(0, curr);           
        }

        static bool HandFacingHead(Handed handed)
        {
            Hand hand = Input.Hand(handed);
            if (!hand.IsTracked)
                return false;

            Vec3 palmDirection = (hand.palm.Forward).Normalized;
            Vec3 directionToHead = (Input.Head.position - hand.palm.position).Normalized;

            return Vec3.Dot(palmDirection, directionToHead) > 0.5f;
        }

        public static void DrawHandMenu(Handed handed)
        {
            if (!HandFacingHead(handed))
                return;

            // Decide the size and offset of the menu
            Vec2 size = new Vec2(4, 16);
            float offset = handed == Handed.Left ? -2 - size.x : 2 + size.x;

            // Position the menu relative to the side of the hand
            Hand hand = Input.Hand(handed);
            Vec3 at = hand[FingerId.Little, JointId.KnuckleMajor].position;
            Vec3 down = hand[FingerId.Little, JointId.Root].position;
            Vec3 across = hand[FingerId.Index, JointId.KnuckleMajor].position;

            Pose menuPose = new Pose(
                at,
                Quat.LookAt(at, across, at - down) * Quat.FromAngles(0, handed == Handed.Left ? 90 : -90, 0));
            menuPose.position += menuPose.Right * offset * U.cm;
            menuPose.position += menuPose.Up * (size.y / 2) * U.cm;

            // And make a menu!
            UI.WindowBegin("HandMenu", ref menuPose, size * U.cm, UIWin.Empty);
            if (UI.Button("New Cube"))
            {
                //Model cube = Model.FromMesh(
                //    Mesh.GenerateRoundedCube(Vec3.One * 0.1f, 0.02f),
                //    Default.MaterialUI);

                poseCache.Add(new Pose(0, 0, -0.5f, Quat.Identity));

                //models.Add(cube);
            }

            if (UI.Button("Exit"))
            {
                SK.Quit();
            }

            UI.WindowEnd();
        }
    }
}
`

**Comment by paulmelis** [comment, marked as answer]
The issue is that your use of `UI.Handle("Cube", ref curr, model.Bounds);` means only `poseCache[0]` is every altered. You will need to move the `UI.Handle()` call to within the loop in which you draw the models. In effect, you need a handle for each model, referencing a different `Pose` with each `UI.Handle()` call

**Reply by leeenglestone** [reply]
When I try the following, the cube just snaps back into its initial position once released..


for (int x = 0; x < poseCache.Count; x++)
{                
    //model = models[x];
    Pose pose = poseCache[x];
    UI.Handle("Cube", ref pose, model.Bounds);
    model.Draw(pose.ToMatrix());
}

**Reply by paulmelis** [reply]
Yeah, this is probably because the `"Cube"` you pass needs to be a unique string per handle, see https://stereokit.net/Pages/Reference/UI/Handle.html

**Reply by leeenglestone** [reply]
okay, using this unique Id, I can kind of see how there is multiple cubes now, but they keep snapping back to the original location when released..

for (int x = 0; x < poseCache.Count; x++)
{                
    //model = models[x];
    Pose pose = poseCache[x];
    UI.Handle($"Cube{x}", ref pose, model.Bounds);
    model.Draw(pose.ToMatrix());
}

**Reply by leeenglestone** [reply]
And this demo code isn't using a unique id?

https://github.com/StereoKit/StereoKit/blob/master/Examples/StereoKitTest/Demos/DemoManyObjects.cs

**Reply by paulmelis** [reply]
But there's only one `Handle` in that example, always referencing the same `Pose`.

I guess one issue could be that `Pose` is a struct, which is copied by value. So your `Pose pose = poseCache[x];` makes a copy, which then gets updated by `UI.Handle()`, you then draw with that updated value (only a small change in a single frame), but the entry in `poseCache` never gets updated. Perhaps try `ref poseCache[x]`?

I actually always use the Begin/End form for `Handle`, but it should not behave differently:

```
for (int x = 0; x < poseCache.Count; x++)
{
    Pose pose = poseCache[x];
    UI.HandleBegin($"Cube{x}", ref pose, model.Bounds);
    // Within the Begin/End block the pose is applied already, so don't need it when drawing the model
    model.Draw(Matrix.Identity);
    UI.HandleEnd();
}
```

Edit: btw, I'm merely making suggestions, not actually trying the code as I haven't got VS nearby

**Reply by leeenglestone** [reply]
You cannot use ref and an indexer apparently..

Have tried the above and it still snaps back i'm afraid.

I appreciate your help. I assumed this use case would be a common occurrence..

**Reply by paulmelis** [reply]
> You cannot use ref and an indexer apparently..

That's because `poseCache` is a `List` and not an array.

>Have tried the above and it still snaps back i'm afraid.
>I appreciate your help. I assumed this use case would be a common occurrence..

Okay, that's it, I fired up Visual Studio :)

This works for me, no snapping back:

```
using StereoKit;
using System;
using System.Collections.Generic;

namespace lee
{
    class Program
    {
        static void Main(string[] args)
        {
            // Initialize StereoKit
            SKSettings settings = new SKSettings
            {
                appName = "lee",
                assetsFolder = "Assets",
            };
            if (!SK.Initialize(settings))
                Environment.Exit(1);


            // Create assets used by the app
            Pose cubePose = new Pose(0, 0, -0.5f, Quat.Identity);
            Model cube = Model.FromMesh(
                Mesh.GenerateRoundedCube(Vec3.One * 0.1f, 0.02f),
                Default.MaterialUI);

            Matrix floorTransform = Matrix.TS(0, -1.5f, 0, new Vec3(30, 0.1f, 30));
            Material floorMaterial = new Material(Shader.FromFile("floor.hlsl"));
            floorMaterial.Transparency = Transparency.Blend;

            List<Pose> poseCache = new List<Pose>();
            poseCache.Add(new Pose(0, 0, -0.5f, Quat.Identity));
            poseCache.Add(new Pose(0.5f, 0, -0.5f, Quat.Identity));
            poseCache.Add(new Pose(0, 0.5f, -0.5f, Quat.Identity));

            // Core application loop
            while (SK.Step(() =>
            {
                if (SK.System.displayType == Display.Opaque)
                    Default.MeshCube.Draw(floorMaterial, floorTransform);

                for (int x = 0; x < poseCache.Count; x++)
                {
                    Pose pose = poseCache[x];
                    UI.Handle($"Cube{x}", ref pose, cube.Bounds);
                    cube.Draw(pose.ToMatrix());
                    poseCache[x] = pose;
                }

            })) ;
            SK.Shutdown();
        }
    }
}
```

Edit: updated not to use HandleBegin/End
Edit 2: the unique Handle ID really is necessary, you get funky behaviour otherwise

**Reply by leeenglestone** [reply]
yes sir! that has worked!

This is the very important line I was missing..

`poseCache[x] = pose;`

One thousand thank you's!

**Reply by paulmelis** [reply]
> yes sir! that has worked!
> This is the very important line I was missing..

Great! Yeah, C# struct (pass by value) versus class (pass by reference) is a bit weird :)
